### PR TITLE
Enable Clang sanitizers for debug builds

### PR DESCRIPTION
**Pull request by steveharter**
This enables the Clang sanitizers to the coreclr (https://github.com/dotnet/corefx/issues/4434) which instruments the binaries to detect issues with native code such as referencing invalid memory, or performing 'undefined' behavior like shifting a 32-bit int >= 32 or <0 bits.

When enabling the sanitizers various cryptic (and hard-to-fix) linker errors occur because various classes have functions with missing implementations. The majority of the code changes add #ifdefs to .h files to align to the #ifdefs in the corresponding .cpp files especially around the defines CROSSGEN_COMPILE and DACCESS_COMPILE. I did what I could to enable function-level linking in Clang in order to remove dead code, but the various linker errors still existed when the sanitizers are enabled.

The sanitizers are definitely finding various issues so it’s a good thing. The next step after this is in is to fix a few of the issues to reduce the error output noise, and then log issues for the remaining. Eventually this would be hooked up to the build system in some way.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -253,23 +252,55 @@ if (CLR_CMAKE_PLATFORM_UNIX)
...
+        set(CLR_SANITIZE_LINK_FLAGS "${CLR_SANITIZE_LINK_FLAGS}undefined")
+        message("Undefined Behavior Sanitizer (ubsan) enabled")
+      endif ()
+
+      set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} ${CLR_SANITIZE_CXX_FLAGS}")
+
+      set(CMAKE_EXE_LINKER_FLAGS_DEBUG "${CMAKE_EXE_LINKER_FLAGS_DEBUG} ${CLR_SANITIZE_LINK_FLAGS}")
+
+      # -Wl and --gc-sections: drop unused sections\functions (similar to Windows /Gy function-level-linking)
+      set(CMAKE_SHARED_LINKER_FLAGS_DEBUG "${CMAKE_SHARED_LINKER_FLAGS_DEBUG} ${CLR_SANITIZE_LINK_FLAGS} -Wl,--gc-sections")
```
Comment (janvorli) [inline CMakeLists.txt]: I am not sure this is correct. When linking libcoreclrpal.a into e.g. libcoreclr.so, we explicitly include everything (I had to add a linker option to do so in the past), otherwise linker could remove functions that are pinvoked from mscorlib, but not called from anywhere else.

File: CMakeLists.txt [hunk 2]
```diff
@@ -253,23 +252,55 @@ if (CLR_CMAKE_PLATFORM_UNIX)
...
+    # obtain settings from running enablesanitizers.sh
+    string(FIND "$ENV{DEBUG_SANITIZERS}" "asan" __ASAN_POS)
+    string(FIND "$ENV{DEBUG_SANITIZERS}" "ubsan" __UBSAN_POS)
+    if ((${__ASAN_POS} GREATER -1) OR (${__UBSAN_POS} GREATER -1))
+      # set compiler flags
+      # -fdata-sections -ffunction-sections: each function has own section instead of one per .o file (needed for -Wl and --gc-sections)
+      # -O1: basic level of optimization
+      # -fPIC: required for shared libraries (Position Independent Code)
+      # -fno-omit-frame-pointer: better stack dumps for debugging
+      set(CLR_SANITIZE_CXX_FLAGS "${CLR_SANITIZE_CXX_FLAGS} -fdata-sections -ffunction-sections -O1 -fPIC -fno-omit-frame-pointer -fsanitize=")
```
Comment (janvorli) [inline CMakeLists.txt]: Are you sure we need to set the -fPIC option globally? We set it in the specific CMakeLists.txt for libraries where it is needed, but not everywhere. 
Also, the -fno-omit-frame-pointer is already set further in this CMakeLists.txt (both for release and debug builds), but it is true it is not set for the build of stuff that's not using PAL, inclusing PAL itself (all the stuff around line 292 below), so it might be good to move that setting up and keeping the option for both release and debug builds.
Also, it is not clear to me why would the -fPIC /  -fno-omit-frame-pointer be in the CLR_SANITIZE_CXX_FLAGS when they are not specific to the sanitize.

**Comment by janvorli** [comment]
@steveharter Can you please add description to the PR? To describe what is the motivation for this change.

File: enablesanitizers.sh [hunk 3]
```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env bash
```
Comment (janvorli) [inline enablesanitizers.sh]: Isn't enabling sanitizers a build process thing? What is the purpose of setting env variables and then consuming them in the build instead of adding these options to the build.sh itself?

File: CMakeLists.txt [hunk 4]
```diff
@@ -408,7 +439,7 @@ if(CLR_CMAKE_PLATFORM_UNIX)
 add_definitions(-DDISABLE_CONTRACTS)
 # The -ferror-limit is helpful during the porting, it makes sure the compiler doesn't stop
 # after hitting just about 20 errors.
-add_compile_options(-ferror-limit=4096)
+add_compile_options(-ferror-limit=40)
```
Comment (janvorli) [inline CMakeLists.txt]: Please do not lower this limit. It is better to see as much errors as possible during the build.

File: CMakeLists.txt [hunk 5]
```diff
@@ -253,23 +252,55 @@ if (CLR_CMAKE_PLATFORM_UNIX)
...
 
+  # set the CLANG sanitizer flags for debug build
+  if(UPPERCASE_CMAKE_BUILD_TYPE STREQUAL DEBUG)
+    # obtain settings from running enablesanitizers.sh
+    string(FIND "$ENV{DEBUG_SANITIZERS}" "asan" __ASAN_POS)
+    string(FIND "$ENV{DEBUG_SANITIZERS}" "ubsan" __UBSAN_POS)
+    if ((${__ASAN_POS} GREATER -1) OR (${__UBSAN_POS} GREATER -1))
+      # set compiler flags
+      # -fdata-sections -ffunction-sections: each function has own section instead of one per .o file (needed for -Wl and --gc-sections)
+      # -O1: basic level of optimization
```
Comment (janvorli) [inline CMakeLists.txt]: The optimization flag should not be set here. This flags is set in `src\pal\tools\clang-compiler-override.txt` to -O0. Would the sanitizer have problem with it? It is also not clear to me which one would the compiler use if you add the -O1 here.

File: src/utilcode/clrhost_nodependencies.cpp [hunk 6]
```diff
@@ -383,6 +394,10 @@ FastFreeInProcessHeapFunc __ClrFreeInProcessHeap = (FastFreeInProcessHeapFunc) C
 
 const NoThrow nothrow = { 0 };
 
+#ifdef HAS_ADDRESS_SANITIZER
+// use standard heap functions for address santizier
```
Comment (janvorli) [inline src/utilcode/clrhost_nodependencies.cpp]: It seems that using standard heap functions would change the behavior of the new. Our new throws exception derived from our Exception type and the runtime can handle it at some points. The default new throws a completely different exception.

File: src/utilcode/clrhost_nodependencies.cpp [hunk 7]
```diff
@@ -431,11 +446,17 @@ operator new[](size_t n)
...
+#endif // __has_feature && __has_feature(address_sanitizer)
+
 #ifdef __llvm__
 __attribute__((visibility("hidden")))
 #endif
 void * __cdecl operator new(size_t n, const NoThrow&)
 {
+#ifdef HAS_ADDRESS_SANITIZER
+    // use standard heap functions for address santizier (which doesn't provide for NoThrow)
+	void * result = operator new(n);
```
Comment (janvorli) [inline src/utilcode/clrhost_nodependencies.cpp]: And here we are making version of new that should not throw a throwing one.

File: src/utilcode/clrhost_nodependencies.cpp [hunk 8]
```diff
@@ -454,7 +476,11 @@ __attribute__((visibility("hidden")))
 #endif
 void * __cdecl operator new[](size_t n, const NoThrow&)
 {
-    STATIC_CONTRACT_NOTHROW;
+#ifdef HAS_ADDRESS_SANITIZER
+    // use standard heap functions for address santizier (which doesn't provide for NoThrow)
+	void * result = operator new[](n);
```
Comment (janvorli) [inline src/utilcode/clrhost_nodependencies.cpp]: The same here

File: src/vm/ceeload.h [hunk 9]
```diff
@@ -3686,6 +3686,7 @@ class ReflectionModule : public Module
     // If true, then only other transient modules can depend on this module.
     bool m_fIsTransient;
 
+#if !defined DACCESS_COMPILE && !defined CROSSGEN_COMPILE
```
Comment (janvorli) [inline src/vm/ceeload.h]: What is the reason for this change?

Comment (janvorli) [inline src/vm/ceeload.h]: And all the other similar changes in the ifdefs below?

File: src/vm/compilestubs.cpp [hunk 10]
```diff
@@ -0,0 +1,37 @@
+//
```
Comment (janvorli) [inline src/vm/compilestubs.cpp]: Why do we need these stubs?

File: enablesanitizers.sh (hunk 3 above)
Comment (steveharter) [inline enablesanitizers.sh]: It is both a build and run-time. I decided to create one .sh to do both, instead of build option + some other .sh or documentation. I can't set the run-time env variables in the build.sh as they wouldn't be present in the environment after the build.sh is finished. The run-time variables used by the sanitizers include ASAN_OPTIONS (which is currently used to turn on run-time memory leak detection)  and ASAN_SYMBOLIZERPATH (which is used to resolve the call stack function names)

File: CMakeLists.txt (hunk 4 above)
Comment (steveharter) [inline CMakeLists.txt]: Sure I will remove this. It was useful for me when I only cared about the first few errors, as the rest were just derivatives.

File: CMakeLists.txt (hunk 1 above)
Comment (steveharter) [inline CMakeLists.txt]: I'll add --no-gc-sections to src\pal\src\CMakeLists.txt which should correct any issues

File: enablesanitizers.sh (hunk 3 above)
Comment (janvorli) [inline enablesanitizers.sh]: Ah, makes sense then.

File: CMakeLists.txt (hunk 1 above)
Comment (janvorli) [inline CMakeLists.txt]: Is this option necessary for the sanitizer?

File: src/utilcode/clrhost_nodependencies.cpp (hunk 6 above)
Comment (steveharter) [inline src/utilcode/clrhost_nodependencies.cpp]: Yes good point. This is an issue when using the sanitizer as it doesn't support the modes that throw an exception. Not sure of the impact (tests?) yet as I need to fix some other issues first. One option is to wrap these and throw an exception ourselves.

File: CMakeLists.txt (hunk 5 above)
Comment (steveharter) [inline CMakeLists.txt]: Interestingly,  -O0 does cause compile errors (I didn't dig into that yet). -O1 also helps with dead code elimination so that was the main reason. Does pal require -O0 (in debug build?)

File: CMakeLists.txt (hunk 1 above)
Comment (jkotas) [inline CMakeLists.txt]: > otherwise linker could remove functions that are pinvoked from mscorlib, but not called from anywhere else. 

This should not be a problem anymore now that we have explicit control over libcoreclr.so exports.

File: CMakeLists.txt (hunk 5 above)
Comment (janvorli) [inline CMakeLists.txt]: The point to use -O0 was to make sure the debug build is easily debuggable without surprises caused by optimizations.

File: CMakeLists.txt (hunk 1 above)
Comment (janvorli) [inline CMakeLists.txt]: Unfortunatelly, it is not the case. Even if you add a function as an export, the linker can happily eliminate it if it comes from a static library. I've bumped into this problem in the past, that was the primary reason why the hosting APIs in the past when they were just a thin wrapper around the COM ones needed to be in a separate unixinterface.cpp.

File: src/vm/compilestubs.cpp (hunk 10 above)
Comment (jkotas) [inline src/vm/compilestubs.cpp]: I have submitted PR https://github.com/dotnet/coreclr/pull/2024 that should make these stubs unnecessary.

File: src/vm/ceeload.h (hunk 9 above)
Comment (jkotas) [inline src/vm/ceeload.h]: The sanitizer is doing de-optimizations that are introducing references to unreachable code - that is not there in some build configs (dac, crossgen).

File: src/vm/compilestubs.cpp (hunk 10 above)
Comment (jkotas) [inline src/vm/compilestubs.cpp]: (I believe similar approach can be used to make the ifdefs around EEJitManager cleaner - ifdef out the whole class.)

File: CMakeLists.txt (hunk 1 above)
Comment (steveharter) [inline CMakeLists.txt]: Adding --gc-sections helps remove dead code, which reduces the number of unresolved symbols added to the libraries (from within that dead code), which prevents at least some linker errors when compiling under the sanitizer.
IMO we should look at add --gc-sections even to release build, as it reduces the size of the library (along with -fdata-sections -ffunction-sections)

File: src/vm/ceeload.h (hunk 9 above)
Comment (steveharter) [inline src/vm/ceeload.h]: Correct on the de-optimization. Most of these changes make the .h file match the .cpp with respect to #ifdefs. For example, ceeload.h defines ReflectionModule::IsMetadataCaptureSuppressed() but the .cpp has #ifdefs for DACCESS_COMPILE and CROSSGEN_COMPILE. So the change shown above adds the same #ifdef to the .h.

Comment (janvorli) [inline src/vm/ceeload.h]: Ok, makes sense.

Comment (steveharter) [inline src/vm/ceeload.h]: It is possible some of the changes are no longer necessary because I added the gc-sections support for removing dead code after some of the code changes. I'll go threw a few of them and be sure, although in general it's a good idea anyway to keep the #ifdefs in sync IMO

File: CMakeLists.txt (hunk 5 above)
Comment (steveharter) [inline CMakeLists.txt]: The compile errors are due to a bug with sanitizer with -O0 which is resolved with -O1. The issue occurs in the pal area, so it is required to compile with -O1 (just when sanitizer is enabled, not for normal debug build)

File: CMakeLists.txt (hunk 2 above)
Comment (steveharter) [inline CMakeLists.txt]: On fPIC, sanitizer seems to mark some data sections relocatable with static libs so fPIC is required. The linker specifically states to recompile with fPIC. Instead of adding fPIC to each of the CMakeLists.txt, I added it globally. I don't think it has any negative side effects.

Comment (steveharter) [inline CMakeLists.txt]: On  -fno-omit-frame-pointer, I moved that above as suggested and removed the lower one.

Comment (steveharter) [inline CMakeLists.txt]: On CLR_SANITIZE_CXX_FLAGS, it is now clean with just the sanitizer flags

File: src/utilcode/clrhost_nodependencies.cpp (hunk 6 above)
Comment (janvorli) [inline src/utilcode/clrhost_nodependencies.cpp]: Now I actually wonder if we cannot keep the "new" operators unchanged. The reason is that they will end up calling malloc and so the sanitizer should be able to handle it at that level, I guess.
See this stack trace:
```
#0  __GI___libc_malloc (bytes=140737488347536) at malloc.c:2876
#1  0x00007ffff66d7548 in CorUnix::InternalMalloc (szSize=72) at /home/janvorli/git/src/src/pal/src/cruntime/malloc.cpp:114
#2  0x00007ffff66d74de in PAL_malloc (szSize=72) at /home/janvorli/git/src/src/pal/src/cruntime/malloc.cpp:98
#3  0x00007ffff675a674 in HeapAlloc (hHeap=0x1020304, dwFlags=0, numberOfBytes=72) at /home/janvorli/git/src/src/pal/src/memory/heap.cpp:242
#4  0x00007ffff5e721d2 in EEHeapAlloc (hHeap=0x1020304, dwFlags=0, dwBytes=56) at /home/janvorli/git/src/src/vm/hosting.cpp:593
#5  0x00007ffff5e72343 in EEHeapAllocInProcessHeap (dwFlags=0, dwBytes=56) at /home/janvorli/git/src/src/vm/hosting.cpp:635
#6  0x00007ffff5c019da in ClrAllocInProcessHeapBootstrap (dwFlags=0, dwBytes=56) at /home/janvorli/git/src/src/utilcode/clrhost_nodependencies.cpp:351
#7  0x00007ffff5c02c5f in ClrAllocInProcessHeap (dwFlags=0, dwBytes=...) at /home/janvorli/git/src/src/inc/clrhost.h:212
#8  0x00007ffff5c01df2 in operator new (n=56) at /home/janvorli/git/src/src/utilcode/clrhost_nodependencies.cpp:447
```

Comment (steveharter) [inline src/utilcode/clrhost_nodependencies.cpp]: If we remove the #ifdefs and leave the operators unchanged, there will be duplicate signatures for new, delete, etc during the final link of corerun and other executables as the Clang sanitizer lib defines those.

Comment (jkotas) [inline src/utilcode/clrhost_nodependencies.cpp]: I think it is ok as long as it applies to sanitizer builds only ... I suspect that the sanitizer depends on its own implementations of these operators to be used.

File: src/utilcode/clrhost_nodependencies.cpp [hunk 11]
```diff
@@ -17,6 +17,12 @@
 #include "contract.h"
 #include "tls.h"
 
+#if defined PLATFORM_UNIX
```
Comment (jkotas) [inline src/utilcode/clrhost_nodependencies.cpp]: This should rather be #ifdef __llvm__ . You are taking advantage of llvm-specific feature, not Unix-specific feature.

File: src/vm/compilestubs.cpp (hunk 10 above)
Comment (steveharter) [inline src/vm/compilestubs.cpp]: #2024 did make the stubs unnecessary. That file is now removed

File: src/vm/compile.h [hunk 12]
```diff
@@ -801,8 +800,10 @@ class CompilationDomain : public AppDomain,
 
     CQuickArray<RefCache*> m_rRefCaches;
 
-    HRESULT AddDependencyEntry(PEAssembly *pFile, mdAssemblyRef ref,mdAssemblyRef def);
+#ifdef CROSSGEN_COMPILE
```
Comment (jkotas) [inline src/vm/compile.h]: Undo this ifdef (it would cause build break on desktop)

File: src/vm/compile.h [hunk 13]
```diff
@@ -822,6 +823,7 @@ class CompilationDomain : public AppDomain,
     ~CompilationDomain();
 #endif
 
+#ifdef CROSSGEN_COMPILE
```
Comment (jkotas) [inline src/vm/compile.h]: Undo this ifdef

File: src/utilcode/clrhost_nodependencies.cpp (hunk 11 above)
Comment (steveharter) [inline src/utilcode/clrhost_nodependencies.cpp]: done

File: src/vm/compile.h (hunk 12 above)
Comment (steveharter) [inline src/vm/compile.h]: Done. No longer needed

File: src/vm/compile.h (hunk 13 above)
Comment (steveharter) [inline src/vm/compile.h]: Done. No longer needed

File: src/utilcode/clrhost_nodependencies.cpp [hunk 14]
```diff
@@ -454,7 +476,11 @@ __attribute__((visibility("hidden")))
 #endif
 void * __cdecl operator new[](size_t n, const NoThrow&)
 {
-    STATIC_CONTRACT_NOTHROW;
+#ifdef HAS_ADDRESS_SANITIZER
+    // use standard heap functions for address santizier (which doesn't provide for NoThrow)
+	void * result = operator new[](n);
+#else
+	STATIC_CONTRACT_NOTHROW;
```
Comment (jkotas) [inline src/utilcode/clrhost_nodependencies.cpp]: nit - undo white space change

File: src/utilcode/clrhost_nodependencies.cpp [hunk 15]
```diff
@@ -489,7 +519,7 @@ __attribute__((visibility("hidden")))
 void __cdecl
 operator delete[](void *p) NOEXCEPT
 {
-    STATIC_CONTRACT_NOTHROW;
+	STATIC_CONTRACT_NOTHROW;
```
Comment (jkotas) [inline src/utilcode/clrhost_nodependencies.cpp]: Nit -  undo white space change

Comment (steveharter) [inline src/utilcode/clrhost_nodependencies.cpp]: done

File: src/utilcode/clrhost_nodependencies.cpp (hunk 14 above)
Comment (steveharter) [inline src/utilcode/clrhost_nodependencies.cpp]: done

**Comment by steveharter** [comment]
@janvorli @jkotas any other feedback or concerns before merging? I have noted the lack of exception throw behavior. Thanks

**Comment by jkotas** [comment]
LGTM
